openpype/hosts/nuke/plugins/publish/validate_asset_context.py

```diff
@@ -0,0 +1,146 @@
+        failed_instance = get_errored_instances_from_context(context, plugin)
+        if failed_instance:
+            self.log.debug("Attempting selection ...")
+            plugin.select(failed_instance.pop())
+
+
+class ValidateCorrectAssetContext(
+    pyblish.api.InstancePlugin,
+    OptionalPyblishPluginMixin
+):
+    """Validator to check if instance asset context match context asset.
+
+    When working in per-shot style you always publish data in context of
+    current asset (shot). This validator checks if this is so. It is optional
+    so it can be disabled when needed.
+
+    Checking `asset` and `task` keys.
+    """
+    order = ValidateContentsOrder
+    label = "Validate asset context"
+    hosts = ["nuke"]
+    actions = [
+        RepairAction,
+        SelectInvalidNodesAction,
+    ]
+    optional = True
+
+    @classmethod
+    def apply_settings(cls, project_settings):
+        """Apply deprecated settings from project settings.
+        """
+        nuke_publish = project_settings["nuke"]["publish"]
+        if "ValidateCorrectAssetName" not in nuke_publish:
+            return
+
+        deprecated_setting = nuke_publish["ValidateCorrectAssetName"]
+        cls.enabled = deprecated_setting["enabled"]
+        cls.optional = deprecated_setting["optional"]
+        cls.active = deprecated_setting["active"]
+
+    def process(self, instance):
+        if not self.is_active(instance.data):
+            return
+
+        invalid_keys = self.get_invalid(instance, compute=True)
+
+        if not invalid_keys:
+            return
+
+        message_values = {
+            "node_name": instance.data["transientData"]["node"].name(),
+            "correct_values": ", ".join([
+                "{} > {}".format(_key, instance.context.data[_key])
+                for _key in invalid_keys
+            ]),
+            "wrong_values": ", ".join([
+                "{} > {}".format(_key, instance.data.get(_key))
+                for _key in invalid_keys
+            ])
+        }
+
+        msg = (
+            "Instance `{node_name}` has wrong context keys:\n"
+            "Correct: `{correct_values}` | Wrong: `{wrong_values}`").format(
+                **message_values)
+
+        self.log.debug(msg)
+
+        raise PublishXmlValidationError(
+            self, msg, formatting_data=message_values
+        )
+
+    @classmethod
+    def get_invalid(cls, instance, compute=False):
+        """Get invalid keys from instance data and context data."""
+        invalid = instance.data.get("invalid_keys", [])
+
+        if compute:
+            testing_keys = ["asset", "task"]
+            for _key in testing_keys:
+                if _key not in instance.data:
+                    invalid.append(_key)
+                    continue
+                if instance.data[_key] != instance.context.data[_key]:
+                    invalid.append(_key)
+
+        instance.data["invalid_keys"] = invalid
+
+        return invalid
```

Comment:
This function body is fast, actually.. it's very fast. There's really no need for `compute` here - it just adds to complexity. It's actually way more convenient to NOT do that so that if the user clicked "Select invalid" after "Repair" that _then_ it selects nothing - basically indicating that the fix has worked correctly.

Also, here there's no reason to prefix `key` with `_`.

Reply:
that is good suggestion, thank you  ecf144993fb48c0e6d9d8e7c8c0512f805abafe6